[![Edwards Lab](https://img.shields.io/badge/Bioinformatics-EdwardsLab-03A9F4)](https://edwards.flinders.edu.au)
[![DOI](https://www.zenodo.org/badge/146160006.svg)](https://www.zenodo.org/badge/latestdoi/146160006)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
![GitHub language count](https://img.shields.io/github/languages/count/linsalrob/ComputationalGenomicsManual)


# ComputationalGenomicsManual

Robs manual for the computational genomics and bioinformatics class

# About this manual

Rob, Liz Dinsdale, Tom Jeffries, Bruno Gomez-Gil, Jim Mitchell, and several other colleagues and friends have been teaching genomics and metagenomics for a long time. They have written this manual over the course of several years, and in a variety of formats. Rob moved it to markdown using GitHub in Fall 2018 as part of his computational genomics class.

You can view this [manual online](https://linsalrob.github.io/ComputationalGenomicsManual/)

# Companion Videos

Companion videos that accompany this class are available on You Tube on [Rob's YouTube Playlist](https://www.youtube.com/playlist?list=PLpPXw4zFa0uLMHwSZ7DMeLGjIUgo1IBbn).

# Chapter Index

Chapter | Contents
--- | --- | 
1\. | [Linux](Linux/)
2\. | [Conda](Conda/)
3\. | [Python](Python/)
4\. | [Snakemake](Snakemake)
5\. | [Sequencing Overview](Sequencing/)
6\. | [Sequence File Formats](SequenceFileFormats/)
7\. | [Sequence Quality Control](SequenceQC/)
8\. | [Databases](Databases/)
8a. | - [NCBI Edirect](Databases/NCBI_Edirect.md)
8b. | - [NCBI SRA](Databases/SRA.md)
9\. | [Genome Sequencing Overview](GenomeSequencingOverview)
10\. | [Sequence Assembly](SequenceAssembly)
11\. | [ORF Calling](ORFCalling/)
12\. | [tRNA and rRNA identification](tRNA_rRNA/)
13\. | [Annotation Pipelines](AnnotationPipelines/)
14\. | [Metagenomics](Metagenomics/)
15\. | - [Example Data Sets](Metagenomics/ExampleDataSets.md)
16\. | [Cross Assembly](CrossAssembly/)
16a. | - [Metabat](CrossAssembly/Metabat.md)
16b. | - [CCOM](CrossAssembly/CCOM.md)
17\. | [16S sequencing](16S/)
18\. | [Host removal](Deconseq/)
19\. | [FOCUS](FOCUS/)
20\. | [Kraken](Kraken2)
21\. | [SUPER-FOCUS](SUPER-FOCUS/)
22\. | [GenomePeek](GenomePeek/)
23\. | [RTMg](RTMg/)
24\. | [OrfM and the SEED](Metagenomics/AnnotatingOrfMSeed.md)
25\. | [ANVI'O](ANVIO/)
26\. | [CheckM](CheckM/)


# Workshops.

We are using this content in a [variety of workshops](Workshops/)


# Assignments. 

Solutions are still not shown, but you can work through some of these

* [NCBI EDirect](Assignments/NCBIEDirectAssignment) is to familiarize yourself with NCBI EDirect.
* [Genomics Assignment](Assignments/GenomicsAssignment/) is to analyze complete genomes from *Klebsiella*.
* [Metagenomics Assignment](Assignments/MetagenomicsAssignment) is to analyze some metagenomics data and describe the organisms that you find there.


# Datasets

We have several different [datasets](Datasets/) available for you to use to try the course work out. There are both 16S and random metagenomes, and links to genomics data.


# PDFs

Note: The PDFs are automatically created from the markdown, and loose some of the images and links. You should probably use the HTML version most of the time.

# About Copyright Information

Some of the images used in this manual are currently copyright other people. As noted above, Rob and friends wrote this manual over many years and added the images and cartoons to lighten the manual. We are in the process of identifying the copyright holders and/or identifying images that are not copyrighted. If your rights have been infringed upon, if you would like to provide an indemnification, or if you would like to provide a non-copyrighted image, please contact Rob.

# Copyright

This manual is Copyright Robert A. Edwards. 2018.

# Citation

If you wish to cite this manual, please cite: *Edwards, R. 2018. Computational Genomics. https://linsalrob.github.io/ComputationalGenomicsManual/. Accessed [today's date]* DOI: 10.5281/zenodo.7883375

# References

We have an [extensive list of references](References/) available, but if you find something missing that we should have cited (a) we're sorry, we tried to remember all of them and (b) please email Rob or provide a pull request and we'll add it.
